### DISCUSSION TITLE
CGI variable "map" fails to validate.

### DISCUSSION THREAD

**Discussion by MrApe** [camptocamp/docker-mapserver#947] [General]
I'm trying to get the mapserver up and running but get the following error:

`msCGILoadMap(): Web application error. CGI variable "map" fails to validate.`

Here is my compose file:

```
version: '3'
services:
  mapserver:
    image: camptocamp/mapserver
    ports:
      - 8081:80
    volumes:
      - $PWD/mapserver/:/etc/mapserver/
    restart: unless-stopped
```

The local `mapserver` dir contains the following:

```
mapserver
├── data
│   └── states_ugl.shp
└── mapserver.map
```

* `mapserver.map` is the [example file from the map server tutorial](https://mapserver.org/tutorial/example1-1-map.html#example1-1-map).
 * `states_ugl.shp` is the corresponding shapes file.

I changed the `SHAPEPATH` to use the `data`-directory: 

```
SHAPEPATH      "data"
```

<details>

<summary>mapserver.map</summary>

```
# The annotated map file (sort of)
# Created by Pericles S. Nacionales for the MapServer tutorial
# 2005-04-08
# Maintained & enhanced since by Jeff McKenna, GatewayGeo
# 2023-04-19 Last updated
#
# MapServer map file uses the pound sign (#) to denote the start of a line
# comment--each line that needs to be commented has to be prepended with a "#".
#
# Map files begin with MAP keyword to signify the start of the map object.
# Well, the entire map file is THE map object.  Enclosed between MAP and END
# at the very bottom of this map file, are keyword/value pairs and other
# objects.
MAP
  IMAGETYPE      PNG
  EXTENT         -97.238976 41.619778 -82.122902 49.385620
  SIZE           400 300
  SHAPEPATH      "data"
  IMAGECOLOR     255 255 255

  # Layer objects are defined beneath the map object.  You need at least one
  # layer defined in your map file before you can display a map...  You can
  # define as many layers as you'd like.

  # Start of LAYER DEFINITIONS ---------------------------------------------
  LAYER # States polygon layer begins here
    NAME         "states"
    DATA         "states_ugl.shp"
    STATUS       OFF
    TYPE         POLYGON

    # The class object is defined within the layer object.  You can define as
    # many classes as you need (well, there are limits as with layers, but it's
    # senseless to define more than ten on a "normal" layer.  There are
    # situations, however, where you might have to do it.)
    CLASS
      NAME       "The Upper Great Lakes States"

      # There are styles in a class, just like there are classes in a layer,
      # just like there are layers in a map.  You can define multiple styles in
      # a class just as you can define multiple classes in a layer and multiple
      # layers in a map.
      STYLE
        COLOR        232 232 232
        OUTLINECOLOR 32 32 32
      END
    END
  END # States polygon layer ends here
  # End of LAYER DEFINITIONS -------------------------------

END # All map files must come to an end just as all other things must come to..
```
</details>

The query used is:

```
http://localhost:8081/cgi-bin/mapserv?map=/etc/mapserver/mapserver.map
```

What could be the problem?

**Comment by geographika** [comment]
Is there a typo in the querystring?

`http://localhost:8081/cgi-bin/mapserv?map=/etc/mapserver/mapserve.map`

Shouldn't it be `http://localhost:8081/cgi-bin/mapserv?map=/etc/mapserver/mapserver.map`

**Comment by MrApe** [comment]
> Is there a typo in the querystring?
> 
> `http://localhost:8081/cgi-bin/mapserv?map=/etc/mapserver/mapserve.map`
> 
> Shouldn't it be `http://localhost:8081/cgi-bin/mapserv?map=/etc/mapserver/mapserver.map`

Thanks for helping out. It was actually just a type in the issue, not the original query. I double checked and fixed the original post.

**Comment by Zia-** [comment]
Hey Guys,
I am getting similar issue although I exactly followed above steps to spin up the server.

![image](https://github.com/user-attachments/assets/d9a058a9-4460-49b4-9b9a-3509f31a93f8)

When doing curl

`curl -X GET http://localhost:8081/cgi-bin/mapserv?map=/etc/mapserver/mapserver.map
`

![image](https://github.com/user-attachments/assets/af8d4e76-533e-48e4-b404-0a4704670bc9)

Not sure what's going wrong and how to debug it. Just trying to setup a basic minimalistic version of it.
Any kind of help would be highly appreciated.

**Comment by sbrunner** [comment]
And what did you get with this URL `http://localhost:8081/?map=/etc/mapserver/mapserver.map`?

**Comment by Zia-** [comment]
@sbrunner 

`mapserv(): Web application error. Traditional BROWSE mode requires a TEMPLATE in the WEB section, but none was provided.
`
This is what I get.

**Comment by sbrunner** [comment]
MapServer 8 requires a configuration file, did you do that?
https://mapserver.org/mapfile/config.html

**Comment by Zia-** [comment]
@sbrunner : Below is my entire setup:

_docker-compose.yaml:_

```
version: '3'
services:
  mapserver:
    image: camptocamp/mapserver
    ports:
      - 8081:80
    volumes:
      - /mapserver/test/:/etc/mapserver/
    restart: unless-stopped
```
    
 content inside _/mapserver/test_ dir:
 
![image](https://github.com/user-attachments/assets/63a8c4e6-f2ef-472a-8855-ecd234ca46c3)

_mapserver.map_ file:

```
# The annotated map file (sort of)
# Created by Pericles S. Nacionales for the MapServer tutorial
# 2005-04-08
# Maintained & enhanced since by Jeff McKenna, GatewayGeo
# 2023-04-19 Last updated
#
# MapServer map file uses the pound sign (#) to denote the start of a line
# comment--each line that needs to be commented has to be prepended with a "#".
#
# Map files begin with MAP keyword to signify the start of the map object.
# Well, the entire map file is THE map object.  Enclosed between MAP and END
# at the very bottom of this map file, are keyword/value pairs and other
# objects.
MAP
  IMAGETYPE      PNG
  EXTENT         -97.238976 41.619778 -82.122902 49.385620
  SIZE           400 300
  SHAPEPATH      "data"
  IMAGECOLOR     255 255 255

  # Layer objects are defined beneath the map object.  You need at least one
  # layer defined in your map file before you can display a map...  You can
  # define as many layers as you'd like.

  # Start of LAYER DEFINITIONS ---------------------------------------------
  LAYER # States polygon layer begins here
    NAME         "ne_10m_land"
    DATA         "ne_10m_land.shp"
    STATUS       OFF
    TYPE         POLYGON

    # The class object is defined within the layer object.  You can define as
    # many classes as you need (well, there are limits as with layers, but it's
    # senseless to define more than ten on a "normal" layer.  There are
    # situations, however, where you might have to do it.)
    CLASS
      NAME       "The Upper Great Lakes States"

      # There are styles in a class, just like there are classes in a layer,
      # just like there are layers in a map.  You can define multiple styles in
      # a class just as you can define multiple classes in a layer and multiple
      # layers in a map.
      STYLE
        COLOR        232 232 232
        OUTLINECOLOR 32 32 32
      END
    END
  END # States polygon layer ends here
  # End of LAYER DEFINITIONS -------------------------------

END # All map files must come to an end just as all other things must come to...
```

I'm running it using `docker-compose up` command.

This is what I see in my terminal:

![image](https://github.com/user-attachments/assets/de21e6ee-6ddc-4275-9856-71b76685efff)

Is there any other config file that I need to define? Am else I am missing in this minimalistic setup?

Thanks a lot!

**Comment by geographika** [comment]
@Zia- there is a working Docker file using the camptocamp image at https://github.com/geographika/getting-started-with-mapserver/blob/main/workshop/exercises/docker-compose.yml
Docs at https://geographika.github.io/getting-started-with-mapserver/

It also has a config file https://github.com/geographika/getting-started-with-mapserver/blob/main/workshop/exercises/mapfiles/mapserver.conf referenced in the docker-compose file:

```
    environment:
      MAPSERVER_CONFIG_FILE: "/etc/mapserver/mapserver.conf"
```

**Comment by mbabic84** [comment]
Hello all,

I have similar issue, but in my case, I have working configuration based on camptocamp/mapserver:8.0, but when I try to update to version 8.2 or even 8.4. Mapserver will stop working with error "Web application error. CGI variable "map" fails to validate.". The only thing I'm changing is image version, nothing else. Any idea what could change between version 8.0 and 8.2?

Thanks
